driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2735,9 +2735,19 @@ public void handle(Message.Response response) {
         case TOPOLOGY_CHANGE:
           ProtocolEvent.TopologyChange tpc = (ProtocolEvent.TopologyChange) event;
           Host.statesLogger.debug("[{}] received event {}", tpc.node, tpc.change);
-          // Always do a full refresh for topology changes. This is simpler than trying to infer a
-          // new EndPoint from the rpc_address only.
-          submitNodeListRefresh();
+          // Do NOT translate the address, it will be matched against Host.getBroadcastRpcAddress()
+          // to find the target host.
+          switch (tpc.change) {
```

Comment:
This is the logic pre-cloud API.

Reply:
If we do this, we should also re-add the branches for `NEW_NODE` and `REMOVED_NODE` in `NodeRefreshRequestDeliveryCallback`, see [this](https://github.com/datastax/java-driver/blob/a815196b81a6180d555b60b08bcbec2e87673fff/driver-core/src/main/java/com/datastax/driver/core/Cluster.java#L2924) (I don't understand how the test can pass without those).

But then we'll run into the issue that we can't create a new node for ADDED because we don't have the host id (like in the UP case where you added a comment below). This is why I changed to a full refresh in the first place.